endpoint.cpp

```diff
@@ -251,22 +246,48 @@ int Endpoint::read_msg(struct buffer *pbuf, int *target_sysid, int *target_compi
          * Ground Station and Flight Stack. Although it can also be a
          * corrupted message is better forward than silent drop it.
          */
-        if (!_check_crc(msg_entry))
+        if (!_check_crc(msg_entry)) {
+            _stat.read.crc_error++;
+            _stat.read.crc_error_bytes += expected_size;
             return 0;
+        }
     }
 
+    _stat.read.handled++;
+    _stat.read.handled_bytes += expected_size;
+
+    if (!_system_id && (!_crc_check_enabled || msg_entry))
+        _system_id = sysid;
+
+    if (_system_id && _system_id != sysid)
+        log_warning("Different system_id message for endpoint %d: Current: %u Read: %u", fd,
+                    _system_id, sysid);
+
     *target_sysid = -1;
-    *target_compid = -1;
 
     if (msg_entry == nullptr)
         log_error("No message entry for %u", msg_id);
     else {
         if (msg_entry->flags & MAV_MSG_ENTRY_FLAG_HAVE_TARGET_SYSTEM)
             *target_sysid = payload[msg_entry->target_system_ofs];
+    }
 
-        if (msg_entry->flags & MAV_MSG_ENTRY_FLAG_HAVE_TARGET_COMPONENT)
-            *target_compid = payload[msg_entry->target_component_ofs];
+    // Check for sequence drops
+    if (_stat.read.expected_seq != seq) {
+        if (_stat.read.total > 1) {
+            uint8_t diff;
+
+            if (seq > _stat.read.expected_seq)
+                diff = (seq - _stat.read.expected_seq);
+            else
+                diff = (UINT8_MAX - _stat.read.expected_seq) + seq;
+
+            _stat.read.drop_seq_total += diff;
+            _stat.read.total += diff;
+        }
+        _stat.read.expected_seq = seq;
```

Comment:
checking the math here.... I can't see any obvious thing that's wrong.  The final numbers don't sum up though.  Are you sure the other end is sending correct numbers?  Or do you have any explanation?